src/chains/algorand_1/algoCompose.ts

```diff
@@ -48,18 +51,54 @@ const ComposeAction: { [key: string]: (args: any, suggestedParams: AlgorandTxHea
   Payment: PaymentTemplate,
 }
 
+/** compile an uncompiled TEAL program (into a Uint8Array) */
+export async function compileFromSourceCode(sourceCode: string, algoClient: AlgoClient): Promise<Uint8Array> {
+  const encoder = new TextEncoder()
+  const programBytes = encoder.encode(sourceCode)
+  const compileResponse = await algoClient.compile(programBytes).do()
+  return new Uint8Array(Buffer.from(compileResponse.result, 'base64'))
+}
+
+/** compile the TEAL program if needed */
+export async function compileIfSourceCodeAndEncode(
+  program: string | Uint8Array,
+  algoClient: AlgoClient,
+): Promise<string> {
+  if (isHexString(program)) {
+    return program as string
+  }
+  // compile the uncompiled program (into a hex string)
+  const byteCode = await compileFromSourceCode(program as string, algoClient)
+  return byteArrayToHexString(byteCode)
+}
+
+// TODO: composeAction expects source code to be sent for appAprovalProgram & appClearProgram
+// Must check if its not a valid hex before it assumes source code has been passed (isValidHex() will be impelemented)
 /** Compose an object for a chain contract action */
 export async function composeAction(
   chainState: AlgorandChainState,
   chainActionType: ChainActionType | AlgorandChainActionType,
-  args: any,
+  args: AlgorandTxAction | AlgorandTxActionRaw | AlgorandTxActionSdkEncoded,
 ): Promise<AlgorandTxActionSdkEncoded> {
   const composerFunction = ComposeAction[chainActionType as string]
   if (!composerFunction) {
     notSupported(`ComposeAction:${chainActionType}`)
   }
+  const appApprovalProgram = args?.appApprovalProgram
+    ? await compileIfSourceCodeAndEncode(args.appApprovalProgram, chainState.algoClient)
+    : undefined
+  const appClearProgram = args?.appClearProgram
+    ? await compileIfSourceCodeAndEncode(args.appClearProgram, chainState.algoClient)
+    : undefined
+  const appArgs = args?.appArgs // TODO: check if appArgs are valid base64 encoded, if not encode to base64
```

Comment:
Is this still needed? It shouldn't be here anyways - it should be in ActionHelper